### PR TITLE
Refactor CodeQL setup

### PR DESCRIPTION
**Pull request by henrymercer**
This PR refactors how the Action sets up the CodeQL tools.  The aim is to make the code more readable and to set the stage for improving how we roll out new CodeQL versions.

### Merge / deployment checklist

- [x] Confirm this change is backwards compatible with existing workflows.
- [x] Confirm the [readme](https://github.com/github/codeql-action/blob/main/README.md) has been updated if necessary.
- [x] Confirm the [changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md) has been updated if necessary.

### CODE REVIEW

**Review by angelapwen — DISMISSED** [review]
Thank you for this!!!! It's so much clearer now. And, I learned that we have `AssertNever` now

File: src/codeql.ts [hunk 1]
```diff
@@ -403,21 +403,162 @@ async function getCodeQLBundleDownloadURL(
...
+    };
+
+async function getCodeQLSource(
+  toolsInput: string | undefined,
+  bypassToolcache: boolean,
+  apiDetails: api.GitHubApiDetails,
+  variant: util.GitHubVariant,
+  logger: Logger
+): Promise<CodeQLToolsSource> {
+  if (toolsInput && toolsInput !== "latest" && !toolsInput.startsWith("http")) {
```
Comment (aeisenberg) [inline src/codeql.ts:427]: Minor:

```suggestion
  if (toolsInput !== "latest" && !toolsInput?.startsWith("http")) {
```

File: src/codeql.ts [hunk 2]
```diff
@@ -403,21 +403,162 @@ async function getCodeQLBundleDownloadURL(
...
+  // This avoids leaking Enterprise tokens to dotcom.
+  // We also don't want to send an authorization header if there's already a token provided in the URL.
+  if (
+    codeqlURL.startsWith(`${apiDetails.url}/`) &&
+    !searchParams.has("token")
+  ) {
+    logger.debug("Downloading CodeQL bundle with token.");
+    headers.authorization = `token ${apiDetails.auth}`;
+  } else {
+    logger.debug("Downloading CodeQL bundle without token.");
```
Comment (aeisenberg) [inline src/codeql.ts:524]: In this case, the URL may still have a `token` search param, Should we explicitly remove it?

Comment (aeisenberg) [reply src/codeql.ts:524]: Looks like this was always here, so maybe not a problem?

File: src/codeql.ts (hunk 1 above)
Comment (henrymercer) [reply src/codeql.ts:427]: I think that would start accepting `toolsInput === undefined`, which we don't want here.

File: src/codeql.ts (hunk 2 above)
Comment (henrymercer) [reply src/codeql.ts:524]: "without adding token" would be more accurate — the goal here is to log that we aren't using the `token` input to `init` here.

**Review by aeisenberg — APPROVED** [review]

**Pull request merged**